Doco fixes + INL cluster installation instructions

GiudGiud opened the pull request:
~~Not quite ready for review, just putting a draft online~~

~~Definitely will need @loganharbour and @milljm 's help to finish up those instructions~~

GiudGiud committed:
Fix some missed titles for doc pages

GiudGiud committed:
quite -> quiet in setupinterface doco

GiudGiud committed:
Add page with installation instructions for INL clusters
Address Fande's review
Closes #20071

milljm commented on modules/doc/content/getting_started/installation/inl_hpc_install_moose.md:
```diff
@@ -0,0 +1,54 @@
+The [conda-based installation](conda.md) is not recommended for INL HPC, as it does not use
+optimized, architecture-specific compilers. However, if you are not seeking performance,
+and will not use more than one node (to limit the attrition of computational resources),
+you are free to use them.
```
👍

cticenhour replied:
I like this (agreeing with @milljm), but I would suggest making this a yellow warning tip box, so that it stands out.

GiudGiud replied:
ok just updated. 
Will be a lot of boxes on this page but not really a problem

cticenhour replied:
Yeah, I think that's fine if it actually helps clarify

GiudGiud commented on modules/doc/content/getting_started/installation/inl_hpc_install_moose.md:
```diff
@@ -0,0 +1,54 @@
+
+- Lemhi: `module load mvapich2/2.3.6-gcc-11.2.0-w2fi cmake`
+
+- Sawtooth: `module load mvapich2/2.3.5-gcc-8.4.0 cmake`
```
yeah that works. It selects the right versions. thanks!

GiudGiud commented on modules/doc/content/getting_started/installation/inl_hpc_install_moose.md:
```diff
@@ -0,0 +1,54 @@
+
+- Lemhi: `module load PETSc/3.15.1-gcc-11.2.0`
+
+- Sawtooth: Currently unavailable
```
ok so please let me know if you're moving forward with removing the petsc modules. If so, then I ll remove most of this section.

dschwen replied:
This seems resolved for now. If the petsc module goes away we can ammend.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/935232/) on 609c685 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20068/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/935343/) on 3338832 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20068/coverage/xfem)

</details>



This comment will be updated on new commits.

dschwen reviewed (dismissed):
:+1:

fdkong commented on modules/doc/content/getting_started/installation/inl_hpc_install_moose.md:
````diff
@@ -0,0 +1,99 @@
+
+```
+  export CC=mpicc CXX=mpicxx
+```
````
FC=mpif90

fdkong replied:
We really should auto this instead of asking set environment. 

I will do that soo

fdkong reviewed (changes requested):
This is very good doc. We might improve by reorganizing a bit. We might talk about several cases separately instead of mixing them. 

1) Easy case: load the end animal binary

2) Intermediate: Load PETSc, build LibMesh and MOOSE

3) Hard: Load MVAPICH, build PETSc, LibMesh, MOOSE

GiudGiud commented:
@fdkong updated the docs with separate cases

GiudGiud commented:
https://mooseframework.inl.gov/docs/PRs/20068/site/getting_started/installation/inl_hpc_install_moose.html is the website page btw
it s not updated rn but soon should be

fdkong commented:
@milljm @loganharbour 

We might make a decision on how to support PETSc on INL HPC. I am fine to go with either way, but we have to make a decision. 

@GiudGiud described three ways here. We need to make sure we have the accurate docs and the accurate practices

fdkong commented on modules/doc/content/getting_started/installation/inl_hpc_install_moose.md:
```diff
@@ -0,0 +1,138 @@
+In order to compile MOOSE and its applications, we first need to load compilers & compiling tools.
+The following modules are recommended for each of INL's clusters:
+
+- Falcon: `module load use.moose MVAPICH2 CMake`
```
Does use.moose already include PETSc?

milljm replied:
use.moose opens the door to load other modules. Think of it like changing into a directory and performing a new directory listing.

So while use.moose doesn't load a PETSc module, it allows for one to be loaded.

GiudGiud replied:
nope, or at least not on sawtooth

GiudGiud replied:
also if you load petsc without it, you will get an intel compiler one (on at least one of the machines)

fdkong reviewed (dismissed):
Approved it with the assumption you have verified all the options on all the machines

GiudGiud commented:
I have. 
My comment though is that the petsc module on sawtooth is old (3.12.5), so it may be trouble for some users

milljm commented:
I need to spend more time with the modules we have set forth on our HPC machines. I would like to see our instructions instead state something like this:

```bash
module load use.moose moose-dev
```

Simple... and never changing. As far as instructions go. Yet, we can change what moose-dev loads in the future. Especially considering what version of PETSc it loads (if at all).

If you modify your instructions to include the above, I'll begin creating these modules on our HPC systems.

fdkong commented:
>My comment though is that the petsc module on sawtooth is old (3.12.5), so it may be trouble for some users

It was why I pinged Logan and Jason earlier

GiudGiud commented:
ok so no CMake, mvapich2 or PETSc just moose-dev?
this needs to be somewhat set in stone. We cant forget to include petsc in the instructions if we remove it from moose-dev

milljm commented:
Sorry for the delay, been a fun day thus far :)

milljm commented:
We can load any other module inside a module. So I will follow any advice given here. Do we do it in three modules, or two?

```bash
module load use.moose moose-dev
```
(loads everything, including petsc)

```bash
module load use.moose moose-dev PETSc
```
(loads mvapich2, cmake in moose-dev, and loads PETSc in the PETSc module)

GiudGiud commented:
Option 1 if someone keeps the petsc module up to date.
Option 2 if not.

GiudGiud commented:
does it take special rights to make HPC modules? could I do it?

milljm commented:
It does... the modules are owned by a system process. A few of us have access.

milljm commented:
@fdkong we were in another meeting about "should we continue to maintain a built petsc". I think we need @permcody  opinion on it. Whether or not to continue building PETSc (every version from hence forth) as a module.

Honestly that is the issue, internally. Instructional wise its great. But when a user starts to use this 'PETSc' module, and then suddenly we take it away from them for a later version... it will cause issues.

GiudGiud commented:
Historically I have had the update_and_rebuild_petsc script die on me for various reasons (mirrors down, network issues, glitches) numerous times. For the sake of reliability I think we need a PETSc module. But I am not the one building it so I m not the one doing the work..

loganharbour commented:
The problem is that there isn’t simply “a” petsc module. Versions will change. Defaults versions will change. And often.

fdkong commented:
> Like every 6 months roughly? @fdkong

Yes, it is close to what we are doing now, but we could do often. We might not make an assumption here.

One way to move forward is that: we build petsc with postfix such as PETSc-3.12.x, PETSc-3.13.x, PETSc-3.14.x, PETSc-3.15.x, PETSc-3.16.x, etc. 

To certain point, we will start to truncate the old ones

GiudGiud commented:
And we point to the one we need currently with the PETSc(no suffix) module name
The HPC team provides most modules to everyone else? Would they take care of building the PETScs for us?

milljm commented:
We are privileged enough to be allowed an area to create our own modules. So it should fall to us to creating it.

loganharbour commented:
The only way I would approve of this is if we don't provide a default and force the users to pick a version.

GiudGiud commented:
I have no qualms with a PETSc default as long as it works for our apps.

fdkong commented:
I think we should have a default PETSc that should be consistent with submodule PETSc.

We should have something like this as the default: `PETSc` (point to petsc submodule (3.15)).

Tomorrow, I am going to update PETSc to 3.16 in moose, and then our  HPC `PETSc` should point to 3.16, at the same time, PETSc-3.15 module should be created.  In these cases, if users have trouble with new PETSc, they can roll back to PETSc-3.15, and we have time to take  a close look into the potential issues on HPC

loganharbour commented:
No. We cannot do defaults because people on HPC may be running older codes on purpose, and we cannot change that default under them when we update petsc. I feel very strongly about this.

milljm commented:
Yes, the problem is maintaining all these versions of PETSc, we can't have a default. What you are suggesting is exactly what I meant by taking a version of PETSc right out from underneath them:

Today:
```pre
module load use.moose PETSc

Currently Loaded Modules:
  1) use.moose                  4) miniconda
  2) gcc/9.2.0-gcc-4.8.5-bxc7   5) PETSc/3.12.5-GCC
  3) mvapich2/2.3.3-gcc-9.2.0
```

Tomorrow:
```pre
module load use.moose PETSc

Currently Loaded Modules:
  1) use.moose                  4) miniconda
  2) gcc/9.2.0-gcc-4.8.5-bxc7   5) PETSc/3.16.0-GCC
  3) mvapich2/2.3.3-gcc-9.2.0
```

milljm commented:
When I say "problem maintaining all these versions of PETSc" I mean that by something we can solve.
But we shouldn't have a default. I think that will just cause to much of a headache for the developer.

I realize we're doing _defaults_ for a certain other 'milestone' (release binaries). But that is different in the fact no further development is done with those.

fdkong commented:
It is unclear to me how to resolve the inconsistency we are facing right now without a default HPC PETSc.

If users intend to use old PETSc release, they might specify PETSc version such as: module load PETSc-3.13 or so.

Without a default HPC PETSc, old PETSc fans also need to do the the same thing anyway.

At the same time, I understand that users might complain that they get different results with using the same modules because we change the underneath PETSc.  But it should not be a big concern. 1) they have an option to roll back the old ones. 2) We will do a lot of tests when are rolling out new submodule PETSc in MOOSE.

I am missing something else

milljm commented:
I am suggesting we remove a default, and have users specify the version of PETSc they want. So no more of this:

```bash
$> module load use.moose PETSc
```
Users will need to list the available versions and decide what to use (or to build their own).

```bash
$> module load use.moose
$> module avail

---------------------- /apps/moose/modules/modulefiles ----------------------
   PETSc/3.11.4-GCC        gperftools/2.9    moose-singularity
   PETSc/3.12.5-GCC (D)    miniconda         patchelf/0.9-8.5.0
   alternate-stack         moose-apps

$> module load PETSc/3.11.4-GCC
```

fdkong commented:
BTW, I am not feeling strongly on the default PETSc module as long as we have a way to provide the latest  PETSc for users when they want to use. 

We might just keep telling users: you need to load xxx version of PETSc

loganharbour commented:
I could see the addition of a script that we put in moose that loads the correct recommended HPC environment given the version of moose they're on.

fdkong commented:
>I am suggesting we remove a default, and have users specify the version of PETSc they want.

I am fine with that. At least that will be way better than what we have now. I think the benefit of this model is that we do not really disturb users when we are doing our integration

The worst situation is that: we have a default PETSc, but that PETSc is super old.

milljm commented:
I don't see an issue with just providing a working developers environment (that is MPI, GCC, cmake, python). And then users simply build their own PETSc.

```bash
module load use.moose moose-dev
./update_and_rebuild_petsc.sh
```

I can even set MOOSE_JOBS for them when they load `moose-dev`

GiudGiud commented:
It takes forever and it fails super often. I'd rather we have a module.

milljm commented:
I don't understand by it fails all the time though. Because what I will be doing anyway, is loading up a developers environment, and running ./update_and_rebuild_petsc.sh myself.

fdkong commented:
> It takes forever and it fails super often.

I did not see that you reported  anything :-). It was very smooth for me ....

fdkong commented:
I am a big fan of running the script, but I understand  users want to load something, and then just use that module

milljm commented:
We used to create a libmesh module... that got a bit crazy to maintain. I just know that we've been trying to phase out petsc modules ever since we created the `update_and_rebuild_petsc.sh` script.

fdkong commented:
Yes, that was the original plan. It is hard for me to understand why users are willing to run "update_and_rebuild_libmesh.sh" but "update_and_rebuild_petsc.sh"

fdkong commented:
Only five letters are different ... -)

GiudGiud commented:
I dont think that's a good idea. We cant pause research at INL because the parmetis/hypre/other cool library mirror is down

Also,
Having a module cuts our installation time by quite a bit. On my machine (which is much faster than the HPC file system, and likely has better internet, normalized to the number of people downloading stuff)
time MOOSE_JOBS=6 ./update_and_rebuild_petsc.sh   14m14.070s
time MOOSE_JOBS=6 ./update_and_rebuild_libmesh.sh 
time make -j 6

milljm commented:
So do you guys want me to start working on a script which automates the creation of a PETSc module on all the HPC clusters during a push to master?

Keep in mind, this can get pretty wild considering we sometimes change a PETSc configure argument **but not the version**. Which would then also "take a version of PETSc someone was using out from underneath them".

GiudGiud commented:
Is that too often? How often are the HPC binaries updated?

If you do it that often and fully automatically, then let's have a libmesh module too

milljm commented:
To put it bluntly, its very difficult to separate a variant. That is, to allow everyone using the cluster access to each version they may want.

```bash
module load use.moose petsc_32_valgrind_hfd5/3.14.1
module load use.moose petsc_32_valgrind_hfd5/3.15.1
module load use.moose petsc_32_valgrind_hfd5/3.16.0
module load use.moose petsc_32_valgrind/3.14.1
module load use.moose petsc_32_valgrind/3.15.1
module load use.moose petsc_32_valgrind/3.16.0
module load use.moose petsc_32_hdf5/3.14.1
module load use.moose petsc_32_hdf5/3.15.1
module load use.moose petsc_32_hdf5/3.16.0
module load use.moose petsc_64_valgrind_hdf5/3.14.1
module load use.moose petsc_64_valgrind_hfd5/3.15.1
module load use.moose petsc_64_valgrind_hfd5/3.16.0
... etc
```
Now lets talk about building a version of libMesh that is paired with the above PETSc.... for each hpc cluster.... Not saying its impossible. But this will eat up a lot of Civet client time, building all these different versions. Still not sure how I would automate the fact we add a new argument in there (or remove one, say parmetis, or mumps, etc).

GiudGiud commented:
Let's do 64 bits and force hdf5 onto everyone. Same let s include all numerical solvers by default.
If someone wants something custom, they can build it themselves.
I can post on the slack to see if some options need to make the cut

milljm commented:
The issue will still be that user loading:

```bash
module use.moose PETSc/3.16.0
```
Today, may not be the same tomorrow when we add hdf5. And therefore, we would be breaking their 10000 hour run on an HPC machine.

GiudGiud commented:
I think that's unevitable.
We could suffix modules with their creation date and continuously offer the old ones if necessary. Like not remove them very often

milljm commented:
If we suffix them, then the instructions will always be outdated... The only defaults we can safely state on our instruction pages, is to ask the user to build their own PETSc and then libMesh.

GiudGiud commented:
We can make the PETSC/3.16.0 point to the latest date, and put this one with no date on the website

milljm commented:
That could work. We will have to train the user to understand that loading the default today, may not be the same tomorrow, and how to load the previous module.

Are you sure you want to do this? Would you not appreciate the system admin (me) removing a working module out from under you without telling you, other than you finding out yourself by running into a problem half way through your solve on HPC?

GiudGiud commented:
Yeah you are right this isnt desirable.
Ok so instructions will say
``` module spider petsc ```
pick a date
``` module load petsc/3.16.0/<date_of_module>```

GiudGiud commented:
@fdkong given all the petsc options etc, what should be the configuration of the single one we will provide (per petsc version)

milljm commented:
FYI, this sort of automation isn't easy... Logan has been fighting this for over week (and we are still working on it).

Example: we need to know when to _build_ this PETSc module. A date, and or version, or even the petsc submodule hash is not enough (which Logan taught me this morning). We need to identify the PETSc module based on changes to moose/conda/mpich/conda_build_config.yaml.

So the name of the PETSc module will have to include, or be something like this (copying straight from `new_moosebuild`)

```pre
module load moose-petsc/3bce203
```
...is a real petsc module _right now_. What does this mean to you? Because I have no idea. Is it 3.15.1, or 16, or what... Its the hash for conda_build_config.yaml (or maybe for moose/conda/petsc/meta.yaml I forget).

Anyways, just stating for Logan to comment on. And for the rest of us to get a sense of what is going to be involved with automating this properly.

GiudGiud commented:
Not down for a hash.
Let's just build it once every 6 months when Fande updates the petsc submodule. You ll only have to rebuild it when the mvapich2 module (the mpi one we put in moose-dev) changes or if we change configure options halfway in between petsc updates, that's it.

No automation, but we keep the date naming system for the in-between version events.

milljm commented:
yeah me neither. I would rather have automation, then "I know you're doing it manually now... you just need to be better at it" =D

GiudGiud commented:
I think automation isnt helping much here, only bringing update headaches with versioning. We dont need to provide many petscs. Do you want to have a meeting about it? Get everyone s opinions and make a decision

milljm commented:
Yeah should probably have a meeting. I know Fande, Logan, Cody and I spoke about this very thing, briefly.

loganharbour commented:
You cannot create something that users will depend on that can change underneath them. There needs to be management of versions.

GiudGiud commented:
We addressed this question earlier in the conversation.
We ll have users look for a precise date and specify it, not just module load petsc

loganharbour commented:
> We addressed this question earlier in the conversation.
We ll have users look for a precise date and specify it, not just module load petsc

But you just said we don't need multiple modules:
> We dont need to provide many petscs.

?

fdkong commented:
I think we could make a comprise here. We could add a script inside of MOOSE repo. The change of the script will trigger a  a new PETSc module build.

For example, I might gradually update PETSc in MOOSE from 3.16.1, to 3.16.2, ..., 3.16.x. But not all of these updates need to trigger a HPC module build. I might only want to a build HPC module for 3.16.latest.

In other cases, like MPICH change, GCC change, should not trigger a new HPC PETSc module build since these changes do not necessarily  change the compiler stack on HPC. The compiler stack on HPC is pretty stable.

permcody commented:
Don't mind me guys... Clicking approved...

bwspenc commented on modules/doc/content/getting_started/installation/inl_hpc_install_moose.md:
```diff
@@ -0,0 +1,138 @@
+
+- Falcon: `module load use.moose MVAPICH2 CMake`
+
+- Lemhi: `module load use.moose mvapich2 cmake PETSc`
```
Why is `PETSc` in the list for Lemhi and Sawtooth? It's not supplying a compiler, as this section implies, or is it?

GiudGiud replied:
I did that because otherwise it loads the wrong one. Cant remember which one exactly but I can try again

GiudGiud commented:
We need to keep multiple modules because of the scenario in which people are running stuff and we dont want to change the module they are using. 
So any new module we create will not have quite the same name and will not accidentally be swapped in. 

Users will be running 
module load petsc/3.16.0/2022-01-22/fdepoaswfpsdfu because we ll tell them to look. And we will never remove that one.

But also we wont be building petsc modules every day. Once in a while, when needed (someone changed mvapich2 for example)

GiudGiud commented:
I m saying "we" and using "future" tense but really this is what my position is, not necessarily what will be decided tomorrow ofc

GiudGiud committed:
Address feedback from Logan, Jason and Rob:
- add warning about modifying bashrc
- switch to using moose-dev instead of specifying a full suite
- be explicitly specifying a version and a hash when loading petsc, so we dont change the PETSc module behind people s back
Refs #20071

Add more warnigns for compilers possibly changing etc

GiudGiud commented:
New update. Should be even clearer now. 
I still have to test those instructions. I m worried users will end up with the wrong PETSc, as in one that doesnt match. But maybe there s ways to work with module spider to make sure they dont

@milljm I used your moose-dev idea for mvapich2 + cmake. Please let us know when you have created that

@loganharbour insisting on loading a version+hash for petsc should prevent the issues you and Jason mentioned. I added some warnings that providing the same module isnt always possible, and that we wont cater to special requests. I m excited for the automatic workflow that will be able to provide fresh petsc and libmesh regularly still.

@brandon-biggs I reworded the warning a bit, as I changed from bashrc to bash_profile too btw. I have to check that works fine on HPC. Please comment here for anything you'd want in there, or add commits on top of my PR.

GiudGiud commented:
Added more warnings.
I tested the instructions and it worked for me.
@milljm @loganharbour @brandon-biggs this is ready for your feedback or approval at this point.

GiudGiud commented:
Let's wait on new petsc update (and module), the sawtooth module doesnt seem to build for me anymore?
Anyone else hitting that?

```
<<< Found PETSc 3.12.5 installation in /apps/moose/stack/petsc-3.12.5_mvapich2-2.3.3_gcc-9.2.0 ... >>>
checking whether we can compile a trivial PETSc program... no
checking for TAO support via PETSc... no
configure: error: *** PETSc was not found, but --enable-petsc-required was specified.
Running make -j 6...
```

GiudGiud committed:
Remove referencde to petsc module, not convinced it's working anymore, #20071

GiudGiud committed:
Reorder instructions, add OnDemand special notes, use scratch to install the codes
Refs #20071

GiudGiud commented:
ping @loganharbour @milljm @brandon-biggs 
I would like to get this through. If this is not good enough, please help me improve it.

milljm commented:
> @milljm I used your moose-dev idea for mvapich2 + cmake. Please let us know when you have created that

This is available... Forgot to mention when I completed this.

GiudGiud commented:
Yup I use it in the instructions. It works great too

GiudGiud commented:
Thanks Fande

GiudGiud commented:
Refs #20071
